src/main/java/net/snowflake/client/jdbc/SnowflakeChunkDownloader.java

```diff
@@ -352,7 +362,13 @@ private void startNextDownloaders() throws SnowflakeSQLException {
         continue;
       } else {
         // cancel the reserved memory
+        logger.debug("cancel the reserved memory.");
         curMem = currentMemoryUsage.addAndGet(-neededChunkMemory);
+        if (retry > MAX_NUM_OF_RETRY) {
```

Comment:
If we are already halfway through a prefetch (chunksToDownload > 0), then there is a good chance memory will be freed if we wait a bit, but it doesn't make sense to wait endlessly. Wait for MAX_NUM_OF_RETRY and then cancel the prefetch.